crates/forge/src/init.rs

```diff
@@ -0,0 +1,122 @@
+use anyhow::{bail, Result};
+
+use include_dir::{include_dir, Dir, DirEntry};
+use std::ffi::OsStr;
+use std::fs;
+use std::path::{Path, PathBuf};
+
+static TEMPLATE: Dir = include_dir!("starknet_forge_template");
+
+fn create_with_template(dir: &Dir<'_>, base_path: &PathBuf, project_name: &str) -> Result<()> {
+    fs::create_dir_all(base_path)?;
+
+    for entry in dir.entries() {
+        let path = base_path.join(entry.path());
+
+        match entry {
+            DirEntry::Dir(d) => {
+                fs::create_dir_all(&path)?;
+                create_with_template(d, base_path, project_name)?;
+            }
+            DirEntry::File(f) => {
+                let contents = f.contents();
+                let contents = replace_project_name(contents, project_name);
+
+                fs::write(path, contents)?;
+            }
+        }
+    }
+
+    Ok(())
+}
+
+fn replace_project_name(contents: &[u8], project_name: &str) -> Vec<u8> {
+    // SAFETY: We control these files, and we know that they are UTF-8.
+    let contents = unsafe { std::str::from_utf8_unchecked(contents) };
+    let contents = contents.replace("{{ PROJECT_NAME }}", project_name);
+    contents.into_bytes()
+}
+
+fn check_path(path: &Path) -> Result<()> {
+    // warn if the path contains characters that will break `env::join_paths`
+    if std::env::join_paths(std::slice::from_ref(&OsStr::new(path))).is_err() {
+        let path = path.to_string_lossy();
+        bail!(format!(
+            "the path `{path}` contains invalid PATH characters (usually `:`, `;`, or `\"`)\n\
+            It is recommended to use a different name to avoid problems."
+        ));
+    }
+    Ok(())
+}
+
+/// Inspired by scarb package name validation
+/// https://github.com/software-mansion/scarb/blob/main/scarb/src/core/package/name.rs#L57
+fn check_name(name: &str) -> Result<()> {
```

Comment:
Perhaps instead of duplicating this logic, scarb should create a package with utilites like that? We could do a PR for this. 

cc @mkaput

Reply:
That's one of the reasons I suggested that you run `scarb new/init` and then patch what came out of it. Especially as @MrDenkoV is currently changing this logic 😅

Reply:
I am generally unopposed to extract `PackageName` type _somewhere_, but I think it's going to be more painful to maintain, than you relying on these commands

Reply:
My idea was that if scarb has this validation somewhere already, we could just make it a separate crate we could use. Problem with `new/init` is that we have limited control on template structure and patching the outputs is more error prone than just copying.

Reply:
> Problem with new/init is that we have limited control on template structure and patching the outputs is more error prone than just copying.

The trick is that I think the opposite 😆 Primarily, you would have to remember to upgrade this crate whenever stuff changes. On the other hand, we can basically guarantee that `scarb new/init` only produces `src/lib.cairo` with some stuff, and it's `Scarb.toml` won't grow beyond what you have currently; except we may end up adding editions support (in `[package]` section) when they will happen in Cairo. So your patching would actually boil down to appending to `Scarb.toml` and overwriting `src/lib.cairo`. We can huddle to discuss this further.